librad/src/git/fetch.rs

```diff
@@ -30,24 +30,56 @@ use crate::{
     peer::PeerId,
 };
 
+/// Upper bound of 500KB for fetching data. This upper bound is useful for when
+/// fetching from an unknown (and possibly untrusted) peer.
+pub const FIVE_HUNDRED_KB: usize = 1024 * 500;
+/// Relative upper bound of 1KB for fetching data. This bound is used in tandem
+/// with the number references that it will be fetching.
+pub const ONE_KB: usize = 1024;
+
+/// Limits used for guarding against fetching large amounts of data from the
+/// network.
+#[derive(Clone, Copy, Debug)]
+pub struct Limit {
```

Comment:
Hm, the `Replicate` variant needs a much higher limit (like, 3GB, or even 5). I think if we have a peeking limit (regardless of the number of references), a sigrefs limit, and a data limit, it would be easier to reason about it.

Reply:
So I just wanna make sure I got this right, and also follow up with a couple of questions.

We have 3 limits:
1. `peek` that is used in `PeekAll` and `Peek`.
2. `signed_refs` that is used in `PeekAll` -- since we fetch signed refs here too -- and `SignedRefs`
3. `data` that is used in `Replicate`

One question that I have is how should we scale the factor of `peek` and `signed_refs` in the case of `PeekAll`? Do we just have a constant factor of "expect at most 20 remotes"? Should this also be a runtime parameter?

In the case of `data` should it be scaled by the tracked refs and delegates or is it a flat limit, i.e. is it 5GB or 5GB * # of refs? I guess the answer boils down to "Are we potentially fetching 5GB from each of these delegates?", which I don't have a good intuition for.

Reply:
>     2\. `signed_refs` that is used in `PeekAll` -- since we fetch signed refs here too -- and `SignedRefs`

Ah, hm. So why are we fetching `SignedRefs` separately still? :thinking: 

> should we scale the factor

What I'm thinking is that we shouldn't do this at all -- every fetch makes the other side assemble a single packfile, but it is not predictable how large it will be. In the case of `id`-related branches, those should only ever differ by at most a few commits. The same is more or less true for `data` -- say you have a repo with a long history, where fetching the `master` branch would clock at 1GB already. Even if it has a lot of concurrent contributions, it is rather unlikely that those would diverge by gigabytes (or else, something is a bit pathological about that project -- if you still want it, we should provide a knob to raise the limit for that specific repo).

The remaining is `signed_refs`, which is more tricky: every peer has their own, and they may differ substantially between peers. Otoh, every update should be very small, and a single peer's history should delta-compress well. So idk, maybe that'd be a reason to keep fetching `signed_refs` in a separate step, so we can adjust the limit acc. to "real world" numbers.

Reply:
> Ah, hm. So why are we fetching SignedRefs separately still? thinking

Actually, just realising that I added fetching the `singed_refs` to `Peek` and `SignedRefs` isn't actually used :facepalm: I removed this in the latest change.

> if you still want it, we should provide a knob to raise the limit for that specific repo

Right, this makes it more clear to me. Imagining a field where you can jack up the limit and try again.

I set a 5KB limit for peek and 5GB limit for data in the latest change.